File: shopkeeper-angular/src/app/core/services/auth.service.ts
import { Injectable, inject, signal, computed } from '@angular/core';
import {
  Auth,
  GoogleAuthProvider,
  RecaptchaVerifier,
  signInWithPopup,
  signInWithPhoneNumber,
  signOut,
  authState,
  type User,
  type ConfirmationResult,
} from '@angular/fire/auth';
import {
  Firestore,
  doc,
  docData,
  getDoc,
  setDoc,
  serverTimestamp,
} from '@angular/fire/firestore';
import { Observable, of, switchMap } from 'rxjs';
import { toSignal } from '@angular/core/rxjs-interop';

import { UserProfile } from '../models';

/**
 * Authentication + account-profile service.
 *
 * Mirrors the original app's behaviour: on first sign-in a `users/{uid}` profile
 * is created as a plain `user`/`free`/`active` account (admins are promoted in
 * Firestore later); `lastLogin` is stamped on every sign-in. Phone OTP uses
 * Firebase's built-in SMS via invisible reCAPTCHA (no backend required).
 */
@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly auth = inject(Auth);
  private readonly fs = inject(Firestore);

  /** The raw Firebase user (null when signed out). */
  readonly user = toSignal(authState(this.auth), { initialValue: null as User | null });

  /** The signed-in user's account profile, kept live from Firestore. */
  readonly profile = toSignal(this.profile$(), { initialValue: null as UserProfile | null });

  readonly isAuthenticated = computed(() => !!this.user());
  readonly isAdmin = computed(() => this.profile()?.role === 'admin');

  private confirmation: ConfirmationResult | null = null;
  private recaptcha: RecaptchaVerifier | null = null;

  private profile$(): Observable<UserProfile | null> {
    return authState(this.auth).pipe(
      switchMap((u) => {
        if (!u) return of(null);
        return docData(doc(this.fs, 'users', u.uid)) as Observable<UserProfile>;
      }),
    );
  }

  /** Google sign-in (permission is requested by Firebase only on first login). */
  async loginWithGoogle(): Promise<void> {
    const cred = await signInWithPopup(this.auth, new GoogleAuthProvider());
    await this.ensureProfile(cred.user);
  }

  /** Send an OTP by SMS. `containerId` hosts the invisible reCAPTCHA. */
  async sendOtp(phoneE164: string, containerId: string): Promise<void> {
    if (!this.recaptcha) {
      this.recaptcha = new RecaptchaVerifier(this.auth, containerId, { size: 'invisible' });
    }
    this.confirmation = await signInWithPhoneNumber(this.auth, phoneE164, this.recaptcha);
  }

  /** Verify the OTP code; creates/updates the profile on success. */
  async confirmOtp(code: string): Promise<void> {
    if (!this.confirmation) throw new Error('Request an OTP first');
    const cred = await this.confirmation.confirm(code);
    await this.ensureProfile(cred.user);
  }

  async logout(): Promise<void> {
    await signOut(this.auth);
    this.confirmation = null;
  }

  /** Create the profile on first login; stamp lastLogin on subsequent logins. */
  private async ensureProfile(user: User): Promise<void> {
    const ref = doc(this.fs, 'users', user.uid);
    const snap = await getDoc(ref);
    const now = serverTimestamp();
    if (!snap.exists()) {
      const profile = {
        uid: user.uid,
        phone: user.phoneNumber ?? '',
        email: user.email ?? '',
        displayName: user.displayName ?? '',
        provider: user.providerData[0]?.providerId ?? '',
        plan: 'free',
        role: 'user',
        active: true,
        createdAt: now,
        activatedAt: now,
        lastLogin: now,
      };
      await setDoc(ref, profile);
      return;
    }
    await setDoc(ref, { lastLogin: now }, { merge: true });
  }
}

File: shopkeeper-flutter/lib/core/services/auth_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:google_sign_in/google_sign_in.dart';

import '../models/models.dart';

/// Authentication + account-profile provisioning.
///
/// Mirrors the web apps: on first sign-in a `users/{uid}` profile is created as
/// a plain user/free/active account; `lastLogin` is stamped on each sign-in.
/// Google uses native Google Sign-In; phone uses Firebase's SMS OTP.
class AuthService {
  final FirebaseAuth _auth = FirebaseAuth.instance;
  final FirebaseFirestore _db = FirebaseFirestore.instance;

  Stream<User?> authState() => _auth.authStateChanges();
  User? get currentUser => _auth.currentUser;

  Stream<UserProfile?> profile() => _auth.authStateChanges().asyncExpand((u) {
        if (u == null) return Stream<UserProfile?>.value(null);
        return _db.collection('users').doc(u.uid).snapshots().map(
              (d) => d.exists ? UserProfile.fromMap(d.id, d.data()!) : null,
            );
      });

  Future<void> signInWithGoogle() async {
    final googleUser = await GoogleSignIn().signIn();
    if (googleUser == null) throw Exception('Sign-in cancelled');
    final gAuth = await googleUser.authentication;
    final cred = GoogleAuthProvider.credential(
      idToken: gAuth.idToken,
      accessToken: gAuth.accessToken,
    );
    final res = await _auth.signInWithCredential(cred);
    await _ensureProfile(res.user!);
  }

  /// Start phone verification. On Android auto-retrieval may complete sign-in
  /// automatically; otherwise [codeSent] fires with a verificationId.
  Future<void> verifyPhone({
    required String phoneE164,
    required void Function(String verificationId) codeSent,
    required void Function(FirebaseAuthException e) onError,
    void Function()? onAutoVerified,
  }) async {
    await _auth.verifyPhoneNumber(
      phoneNumber: phoneE164,
      verificationCompleted: (cred) async {
        final res = await _auth.signInWithCredential(cred);
        await _ensureProfile(res.user!);
        onAutoVerified?.call();
      },
      verificationFailed: onError,
      codeSent: (id, _) => codeSent(id),
      codeAutoRetrievalTimeout: (_) {},
    );
  }

  Future<void> confirmOtp(String verificationId, String smsCode) async {
    final cred = PhoneAuthProvider.credential(
      verificationId: verificationId,
      smsCode: smsCode,
    );
    final res = await _auth.signInWithCredential(cred);
    await _ensureProfile(res.user!);
  }

  Future<void> signOut() async {
    await GoogleSignIn().signOut();
    await _auth.signOut();
  }

  Future<void> _ensureProfile(User user) async {
    final ref = _db.collection('users').doc(user.uid);
    final snap = await ref.get();
    final now = FieldValue.serverTimestamp();
    if (!snap.exists) {
      await ref.set({
        'uid': user.uid,
        'phone': user.phoneNumber ?? '',
        'email': user.email ?? '',
        'displayName': user.displayName ?? '',
        'provider': user.providerData.isNotEmpty ? user.providerData.first.providerId : '',
        'plan': 'free',
        'role': 'user',
        'active': true,
        'createdAt': now,
        'activatedAt': now,
        'lastLogin': now,
      });
    } else {
      await ref.set({'lastLogin': now}, SetOptions(merge: true));
    }
  }
}
